### ISSUE TITLE
needless_pass_by_ref_mut: false negative, &mut self ignored

### ISSUE THREAD

**Issue by klensy** [rust-lang/rust-clippy#12589]
### Summary

Looks like mutable self refs ignored in methods.

Discovered while working on https://github.com/rust-lang/rust/pull/123188

Checked with clippy from playground 0.1.79 (2024-03-27 c9f8f34)

https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=1eb816feca6371e80e4b27f59b21174f

see also https://github.com/rust-lang/rust-clippy/issues/9591

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
#![warn(clippy::needless_pass_by_ref_mut)]

fn main() {}

struct Foo{
    x: u8
}

impl Foo{
    fn foo(&mut self)->u8{
        self.x * 2
    }

    fn bar(&mut self, y: &mut u8)->u8{
        self.x * *y
    }
    
    fn baz(&self, y: &mut u8)->u8{
        self.x * *y
    }
}

```

I expected to see this happen:
Should warn about unused &mut self in `foo` and `bar`
Instead, this happened:
No warn for &mut self, only for `y`:
```
warning: this argument is a mutable reference, but not used mutably
  --> src/main.rs:14:26
   |
14 |     fn bar(&mut self, y: &mut u8)->u8{
   |                          ^^^^^^^ help: consider changing to: `&u8`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(clippy::needless_pass_by_ref_mut)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: this argument is a mutable reference, but not used mutably
  --> src/main.rs:18:22
   |
18 |     fn baz(&self, y: &mut u8)->u8{
   |                      ^^^^^^^ help: consider changing to: `&u8`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
```

**Comment by klensy** [comment]
Well, self args explicitly ignored why?
https://github.com/rust-lang/rust-clippy/blob/124e68bef8be61aa151ff33bea325c832728146f/clippy_lints/src/needless_pass_by_ref_mut.rs#L85-L87

**Comment by klensy** [comment]
Found discussion at https://github.com/rust-lang/rust-clippy/pull/10900#issuecomment-1591104585, lint description should probably note, that &mut self args currently not linted.

**Comment by GuillaumeGomez** [comment]
I'll add a note for it and a setting to allow to bypass this default behaviour.
